doc/source/dev/howto-docs.rst

```diff
@@ -215,6 +215,218 @@ Note that for documentation within NumPy, it is not necessary to do
 Please use the ``numpydoc`` :ref:`formatting standard <numpydoc:format>` as
 shown in their :ref:`example <numpydoc:example>`.
 
+.. _doc_c_code:
+
+Documenting C/C++ Code
+======================
+
+NumPy uses Doxygen_ to parse specially-formatted C/C++ comment blocks. This generates
+XML files, which are  converted by Breathe_ into RST, which is used by Sphinx.
+
+**It takes three steps to complete the documentation process**:
+
+1. Writing the comment blocks
+-----------------------------
+
+Although there is still no commenting style set to follow, the Javadoc
+is more preferable than the others due to the similarities with the current
+existing non-indexed comment blocks.
+
+.. note::
+   Please see `"Documenting the code" <https://www.doxygen.nl/manual/docblocks.html>`__.
+
+This is how Javadoc style looks like::
+
+  /**
+   * Your brief is here.
+   */
+  void function(void);
+
+For line comment, insert a triple forward slash::
+
+  /// Your brief is here.
```

Comment:
Line comments are in most cases used wrongly. They get added to the `@details` block but out of order (that i. Is there a good use case for not documenting the function only at the toplevel?

```cpp
/**
* @brief This function does something
*  @details it is a good function. The concept behind it is the same as the (ref) algorithm
*/
```

Reply:
Until the first dot, it goes to @brief. Line comments is very useful for C++, see the following example
```C++
/**
 *  Template to represent limbo numbers.
 *
 *  Specializations for integer types that are part of nowhere.
 *  It doesn't support with any real types.
 *
 *  @param Tp Type of the integer. Required to be an integer type.
 *  @param N  Number of elements.
*/
template<typename Tp, std::size_t N>
class DoxyLimbo {
 public:
    /// Default constructor. Initialize nothing.
    DoxyLimbo();
    /// Set Default behavior for copy the limbo.
    DoxyLimbo(const DoxyLimbo &l);
    /// Returns the raw data for the limbo.
    const Tp *data();

    /// @name Comparison operators.
    /// @{

    bool operator==(const DoxyLimbo<Tp, N> &r) const;
    bool operator<(const DoxyLimbo<Tp, N> &r) const;
    bool operator<=(const DoxyLimbo<Tp, N> &r) const;

    /// @}

 protected:
    Tp p_data[N]; ///< Just C array that hold the limbos.
};
```

Reply:
This is a stylistic issue. I personally feel that most new contributors would benefit from the explicit tags, since I do not expect them to know the moires of Doxygen. I also find it more evident (i.e. it is clearer when something is going to be processed by `doxygen`).

Reply:
I don't think documenting C-API interfaces is something we want new contributors to take on until there is a good reference of already-documented interfaces they can copy from. This is the "howto docs" page, we could start another page for a more detailed style guide if needed. I think we should aim for  a bare minimum in this PR so we can get it approved and merged, we can always expand in the future.

Reply:
> This is a stylistic issue

That's why I added the following statement at the beginning:
> Although there is still no commenting style set to follow, the Javadoc
is more preferable than the others due to the similarities with the current
existing non-indexed comment blocks.

And since tags `brief` and `details` aren't part of Javadoc style then we can't require it.